onnx/shape_inference/implementation.cc

```diff
@@ -265,9 +272,15 @@ static void InferShapesImpl(
           existingType = iter->second;
           checkShapesAndTypes(*inferredType, *existingType);
         } else {
-          auto vi = g->add_value_info();
-          vi->set_name(n.output(i));
-          existingType = vi->mutable_type();
+            auto output_iter = outputTypesByName.find(n.output(i));
+            if (output_iter != outputTypesByName.end() && inferredType->tensor_type().elem_type() != TensorProto::UNDEFINED) {
```

Comment:
I am a bit confused by this and the lines below at line number 304: both seem to serve the same purpose. Is there a reason we need both? I think the loop at line 304 should be sufficient, and we can leave this part as before. What do you think?